src/NuGet.Commands/RestoreRequest.cs

```diff
@@ -20,9 +21,35 @@ public RestoreRequest(PackageSpec project, IEnumerable<PackageSource> sources)
         { }
 
         public RestoreRequest(PackageSpec project, IEnumerable<PackageSource> sources, string packagesDirectory)
+            : this(project, packagesDirectory)
         {
+            if (sources == null)
+            {
+                throw new ArgumentNullException(nameof(sources));
+            }
+
+            Sources = sources.Select(source => Repository.Factory.GetCoreV3(source.Source)).ToList();
```

Comment:
GetCoreV3 sets up the V3 resources, which for the project.json resolver includes v2 support since it uses a different provider that does not use nuget.core.